File: backend/app/contracts/alerts.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from backend.app.contracts.vitals import SyntheticProvenance

AlertPriority = Literal["high", "critical"]
DeduplicationStatus = Literal["new_alert", "reused_active", "suppressed_cooldown", "rearmed"]


class AlertResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    alert_id: int
    patient_id: str
    bed_id: str
    priority: AlertPriority
    state: Literal["generated", "assigned", "acknowledged", "responded", "resolved"]
    risk_score: float = Field(ge=0, le=1)
    risk_level: Literal["low", "moderate", "high", "critical"]
    event: str
    probability: float = Field(ge=0, le=1)
    horizon_minutes: int = Field(ge=0)
    observation_sequence: int
    timestamp: datetime
    provenance: SyntheticProvenance
    prototype_label: str
    prediction_source_kind: Literal["ml", "deterministic_fallback"]
    prediction_source_version: str
    fallback_reason: str | None
    prediction_contract_version: str
    effective_threshold: float
    rule_version: str
    deduplication_status: DeduplicationStatus
    created_at: datetime

File: frontend/src/contracts/alerts.ts
import type { OperationalState } from "./realtime";

export type AlertState = "generated" | "assigned" | "acknowledged" | "responded" | "resolved";
export type AlertPriority = "high" | "critical";

export interface AlertEvent {
  event_id: number;
  sequence: number;
  state: AlertState;
  outcome: string;
  occurred_at: string;
}

export interface Alert {
  alert_id: number;
  patient_id: string;
  bed_id: string;
  priority: AlertPriority;
  state: AlertState;
  risk_score: number;
  risk_level: "low" | "moderate" | "high" | "critical";
  event: string;
  probability: number;
  horizon_minutes: number;
  observation_sequence: number;
  timestamp: string;
  provenance: { source_kind: string; source_name: string; scenario_id: string; scenario_version: string; is_live_bedside_feed: boolean };
  prototype_label: string;
  prediction_source_kind: "ml" | "deterministic_fallback";
  prediction_source_version: string;
  fallback_reason: string | null;
  prediction_contract_version: string;
  effective_threshold: number;
  rule_version: string;
  deduplication_status: "new_alert" | "reused_active" | "suppressed_cooldown" | "rearmed";
  created_at: string;
  assignment_id: string | null;
  events: AlertEvent[];
}

export interface AuditEvent {
  audit_id: number;
  sequence: number;
  actor_id: string | null;
  action: string;
  category: string;
  resource_type: string;
  resource_id: string;
  outcome: string;
  resulting_state: string | null;
  correlation_id: string;
  details: Record<string, unknown>;
  occurred_at: string;
}

export type AlertOperationalState = OperationalState | "deterministic_fallback";